fix(python): Create delta compatible schema during writing

ion-elgreco opened the pull request:
Closes #10153  
Closes #10154 

- Casts all uints to ints with the same bitsize
- Cast all timestamps to 'us' precision


Keep getting this type-arg error, but I don't know exactly which dtypes will come from pyArrow, I am just treating it as a black box ( :
`polars/io/delta.py:324: error: Missing type parameters for generic type "list"  [type-arg`]

washcycle commented:
@ion-elgreco does this address the not null schema attributes from my issue #10212 and would your method here be a good place to integrate with? I think the quick fix is to just ignore the nullability. If the schemas are the same and the delta table has columns that cannot be null then pass the schema validation. If the delta target table finds a null in a column delta-rs would throw an exception at that point if the columns did have nulls (need testing here). Polars would boil that exception up. I think polars should ignore nullability for now and it's up to the user to know what they are doing with their deltatable. Otherwise, it might require some hefty modifications to the schema code in this project to take nullability into consideration.

ion-elgreco commented:
> @ion-elgreco does this address the not null schema attributes from my issue #10212 and would your method here be a good place to integrate with? I think the quick fix is to just ignore the nullability. If the schemas are the same and the delta table has columns that cannot be null then pass the schema validation. If the delta target table finds a null in a column delta-rs would throw an exception at that point if the columns did have nulls (need testing here). Polars would boil that exception up. I think polars should ignore nullability for now and it's up to the user to know what they are doing with their deltatable. Otherwise, it might require some hefty modifications to the schema code in this project to take nullability into consideration.

According to Ritchie on discord, Polars doesn't have this nullability concept.

I don't think we would be able to fix this. Since  uint -> int is deterministic but something being nullable you won't know. 

I am also not sure what the impact will be once you write a polars data frame with nulls to a delta table with nullability to false for some columns.

Maybe @ritchie46 has some ideas.

stinodego commented on py-polars/polars/io/delta.py:
```diff
@@ -313,3 +316,109 @@ def _check_if_delta_available() -> None:
+def _reconstruct_field_type(
+    field: pa.Field,
+    field_head: pa.Field,
+    reconstructed_field: list | None = None,
```
You have to specify the type of list here:

```suggestion
    reconstructed_field: list[pa.DataType] | None = None,
```

(not 100% sure this is the correct type, but you get the idea)

stinodego commented on py-polars/polars/io/delta.py:
```diff
@@ -313,3 +316,109 @@ def _check_if_delta_available() -> None:
+                name=field_head.name,
+                type=reduce(lambda x, y: y(x), reversed(reconstructed_field)),
+            )
+    elif "uint" in str(field.type) and not any(
+        isinstance(field.type, dtype) for dtype in [pa.StructType, pa.LargeListType]
+    ):
+        if reconstructed_field is None:
+            return pa.field(
+                name=field.name, type=getattr(pa, str(field.type).strip("u"))()
+            )
+        else:
+            reconstructed_field.append(getattr(pa, str(field.type).strip("u"))())
+            return pa.field(
+                name=field_head.name,
+                type=reduce(lambda x, y: y(x), reversed(reconstructed_field)),
+            )
```
I don't like operating on strings... just define a mapping between types and use that, with some recursive logic to iterate through nested types.

ion-elgreco replied:
While working on something related, maybe we should not do Uint -> Int, since it's possible that a number in Uint is to big to cast into Int. Before I had in my mind that the conversion would be safe, but it's not in some cases.

stinodego replied:
We are not casting, we are just determining what the matching schema would be if it were written to a delta table.

If the value does not fit, the `deltalake` package (which will try to cast it) will throw an error.

ion-elgreco replied:
> We are not casting, we are just determining what the matching schema would be if it were written to a delta table.
> 
> If the value does not fit, the `deltalake` package (which will try to cast it) will throw an error.

Ah you're right. I thought Delta-RS would just silently still write the table.

ion-elgreco replied:
@stinodego regarding your comment. An int, uint, string, largestring are all instance of pa.DataType, from what I can see the only way to distinguish which is which is to check the name of pa.DataType using str(datatype), or datatype.__str__(), or datatype.equals(pa.Uint32)

stinodego reviewed (changes requested):
Can you please resolve any failing tests / lints, and add a test for the bug you're fixing here?

ion-elgreco commented:
> Can you please resolve any failing tests / lints, and add a test for the bug you're fixing here?

Thanks for the feedback! Will do :)

washcycle commented:
> I am also not sure what the impact will be once you write a polars data frame with nulls to a delta table with nullability to false for some columns.

Needs testing, but I think the deltalake (delta-rs) package with handle that. I believe that is built in so when deltal-rs tries to do the actual write it will throw an exception if the data doesn't match the schema. So I propose we just ignore nullability on the polars side and allow deltal-rs to bubble up that those columns cannot have nulls. Does that make sense?

ion-elgreco commented:
> Can you please resolve any failing tests / lints, and add a test for the bug you're fixing here?

So I finished the function. It works when I build the package and try to write the same dataframe as in the test case twice, but somehow it's not working in the test case and I cannot figure out why... 

The weird thing is, `.venv/lib/python3.8/site-packages/pyarrow/dataset.py:935: in write_dataset
    data = InMemoryDataset(data, schema=schema),` it say that the schema it got is not the same as the expected schema. So the new reconstructed schema is passed onto pyarrow through delta-rs,  but the data is still the old schema, so it causes a mismatch. I don't quite understand why this happens during the Pytest, but not during my local execution in a notebook.

If you look at the schema you can see it got reconstructed correctly, by chang Uint to Int and all timestamps to timestamp(us)

ion-elgreco commented:
@stinodego my delta unit test is passing :), But I still have an import regression test that is failing, I don't know how to tackle this. Is it important to solve?

stinodego commented on py-polars/polars/io/delta.py:
```diff
@@ -1,9 +1,12 @@
 from typing import TYPE_CHECKING, Any
 from urllib.parse import urlparse
 
+import pyarrow as pa
```
This causes problems -  you'll have to use our lazy imports:

```suggestion
from polars.dependencies import pyarrow as pa
```

ion-elgreco commented:
@stinodego Tests are all green now 💯

stinodego commented on py-polars/polars/io/delta.py:
```diff
@@ -313,3 +315,114 @@ def _check_if_delta_available() -> None:
         )
+
+
+def _reconstruct_field_type(
```
I believe this currently does not map `LargeString` types yet, does it?

It would be nice to have some dedicated tests for this method to show that it works as intended.

ion-elgreco replied:
Indeed, it skips LargeString because this is catched by the `DeltaTable.schema().to_pyarrow(large_types=True)`

stinodego replied:
If we include it here though (LargeBinary/LargeString/LargeList), the logic becomes way easier to read! We can get rid of a whole bunch of code. And it will make more sense.

ion-elgreco replied:
> If we include it here though (LargeBinary/LargeString/LargeList mapping), the logic becomes way easier to read! We can get rid of a whole bunch of code. And it will make more sense.

Good point! I'll look into creating a helper function that can accommodate also these Large types

ion-elgreco replied:
I've added the large types to normal types mapping.

stinodego commented on py-polars/polars/io/delta.py:
```diff
@@ -313,3 +315,114 @@ def _check_if_delta_available() -> None:
         )
+
+
+def _reconstruct_field_type(
```
Seems to me like this function should really be refactored, lots of nested if-else checks and there seems to be quite some duplication in this pattern. I find it hard to read. A helper function would probably be helpful?

ion-elgreco replied:
I'll have to take a look and rethink the logic. It's quite nested because I started with small statements and kept expanding.

stinodego commented on py-polars/polars/dataframe/frame.py:
```diff
@@ -3405,15 +3409,27 @@ def check_unsupported_types(n: str, t: PolarsDataType | None) -> None:
-            table_schema = table.schema()
+        if data_schema != delta_schema:
+            data_schema = delta_schema
+
+            if table is not None:
+                table_schema = table.schema()
+
+                if data_schema == table_schema.to_pyarrow(as_large_types=True):
+                    data_schema = table_schema.to_pyarrow()
+            else:
+                data = data.cast(data_schema)
+        else:
+            if table is not None:
+                table_schema = table.schema()
 
-            if data_schema == table_schema.to_pyarrow(as_large_types=True):
-                data_schema = table_schema.to_pyarrow()
+                if data_schema == table_schema.to_pyarrow(as_large_types=True):
+                    data_schema = table_schema.to_pyarrow()
```
There must be a better way to write this 🤔

ion-elgreco replied:
> There must be a better way to write this 🤔

I'll have to think about it! Will try to work on it tonight

ion-elgreco replied:
I've removed this part completely now since it became obsolete due to also reconstructing schema from large types to normal types.

stinodego commented on py-polars/tests/unit/io/test_delta.py:
```diff
@@ -179,3 +179,175 @@ def test_write_delta(df: pl.DataFrame, tmp_path: Path) -> None:
+
+    from polars import Datetime, Field, List, Struct, UInt32, Utf8
+
+    df = pl.DataFrame(
+        [
+            pl.Series("string", ["test"], dtype=pl.Utf8),
```
I think these can all be separate test cases instead of one big dataframe. That helps determine which data type fails the test. You can use `@pytest.mark.parametrize` for this.

ion-elgreco replied:
I've parametrized the tests now, each column is a new test case.

stinodego reviewed (changes requested):
I think we're almost there. This will be a nice improvement to the usability of our Delta writer!

I left a few comments mostly on code style. Please try to improve where possible, then I can take it the last step if needed.

ion-elgreco commented:
@stinodego There is a bug in `DataFrame.to_arrow()` when you have a column with a non-UTC timezone. When you do `df.to_arrow()`, it will cast the data to UTC, but still keep the schema in the original TZ.

ion-elgreco commented:
@stinodego I have made all the changes you requested, with regards to refactoring I did my best.

- Added parametrized tests
- Expanded type mapping to Large types
- Removed specific timestamp checks and added them to the type mapping dict
- Changed if else order for improving readability 
- Added helper function to create nested datatype from list

With regards to the rest of `_reconstruct_field_type` I don't see what else I can refactor there, since each specific instance of a type has to return something different

stinodego commented:
Thanks a lot, will take a look at this on Monday and we'll merge it then 👍

ion-elgreco commented:
Just a thought, Should we maybe take care of possible overflow by doing:
```
- Uint8  -> Int16
- Uint16 -> Int32
- Uint32 -> Int64
- Int64  -> Decimal128?
```

stinodego commented:
> Just a thought, Should we maybe take care of possible overflow by doing:
> 
> ```
> - Uint8  -> Int16
> - Uint16 -> Int32
> - Uint32 -> Int64
> - Int64  -> Decimal128?
> ```

No, definitely not! If there is an overflow, there should be an error and the user can cast to the desired types explicitly themselves.

ion-elgreco commented:
> > Just a thought, Should we maybe take care of possible overflow by doing:
> > ```
> > - Uint8  -> Int16
> > - Uint16 -> Int32
> > - Uint32 -> Int64
> > - Int64  -> Decimal128?
> > ```
> 
> No, definitely not! If there is an overflow, there should be an error and the user can cast to the desired types explicitly themselves.

Ok! Currently, the PyArrow library will raise an ArrowInvalidError, not sure if we want to catch it and raise our own error here

stinodego commented:
The PyArrow error is fine for now 👍

ion-elgreco commented:
I've made one last change, previously the original schema was always passed to write_deltalake. Now we only pass the delta-compatible schema if the user didn't provide its custom schema. This is useful if users want to write to delta tables that have nullable columns, I've also added an example for it in the docstring.

It won't solve this https://github.com/pola-rs/polars/issues/10212, but at least the user still maintains the option to override themselves.

stinodego reviewed (approved):
I did quite some refactoring - hope you'll agree that the schema conversion is a lot cleaner now.

Thanks for your effort! This should make Delta writes much more reliable.

Ready to merge! :shipit:

washcycle commented:
> I did quite some refactoring - hope you'll agree that the schema conversion is a lot cleaner now.
> 
> Thanks for your effort! This should make Delta writes much more reliable.
> 
> Ready to merge! :shipit:

Nice! I'm so excited about this one. I wasn't able to use polars because we have a strict delta table schema with columns that cannot be null, but the target is well defined so I can easily get the target schema I am writing and use the schema override option.

stinodego commented:
> > I did quite some refactoring - hope you'll agree that the schema conversion is a lot cleaner now.
> > Thanks for your effort! This should make Delta writes much more reliable.
> > Ready to merge! :shipit:
> 
> Nice! I'm so excited about this one. I wasn't able to use polars because we have a strict delta table schema with columns that cannot be null, but the target is well defined so I can easily get the target schema I am writing and use the schema override option.

Great! Let me know if you run into any trouble.

ion-elgreco commented:
> I did quite some refactoring - hope you'll agree that the schema conversion is a lot cleaner now.
> 
> Thanks for your effort! This should make Delta writes much more reliable.
> 
> Ready to merge! :shipit:

Way cleaner!
